### PR TITLE
WIP: PERF: Cythonize fillna

### LINKED ISSUE #42147: PERF/QST: Why isn't fillna cythonized?
Looking at the fillna code, it looks like a mask is generated and then the fill value is then put in using putmask. Wouldn't it be faster if we skipped the mask generation and just put in nan values directly by looping over in Cython.
https://github.com/pandas-dev/pandas/blob/master/pandas/core/internals/blocks.py#L430-L470

I can take this issue up if this is the correct approach

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #42147
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

Code is still pretty messy at this point, opening mainly for CI purposes. Also, this is my first time messing around with BlockManager so would appreciate a little extra scrutiny on that part.

Benchmarks
<Details>
```

       before           after         ratio
     [9f6a91ae]       [77c05634]
     <master>         <perf-fillna>
-      1.16±0.3ms         772±70μs     0.67  replace.FillNa.time_fillna(True)
-      6.78±0.7ms       4.48±0.3ms     0.66  replace.FillNa.time_fillna_df(False)
-      8.86±0.8ms       2.96±0.2ms     0.33  replace.FillNa.time_fillna_limit(False)
-      7.92±0.6ms       2.46±0.6ms     0.31  replace.FillNa.time_fillna_df(True)
-      6.15±0.5ms         810±60μs     0.13  replace.FillNa.time_fillna_limit(True)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
       before           after         ratio
     [9f6a91ae]       [77c05634]
     <master>         <perf-fillna>
+        644±10μs      1.15±0.04ms     1.79  replace.FillNa.time_replace(True)
+     4.14±0.08ms       5.43±0.3ms     1.31  replace.FillNa.time_replace(False)
-      4.10±0.2ms       2.30±0.2ms     0.56  replace.FillNa.time_fillna(False)
-     1.24±0.09ms         682±20μs     0.55  replace.FillNa.time_fillna(True)
-      6.98±0.4ms       2.61±0.4ms     0.37  replace.FillNa.time_fillna_limit(False)
-      6.61±0.3ms       1.57±0.1ms     0.24  replace.FillNa.time_fillna_df(True)
-      4.99±0.1ms         622±10μs     0.12  replace.FillNa.time_fillna_limit(True)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.

       before           after         ratio
     [9f6a91ae]       [77c05634]
     <master>         <perf-fillna>

-      1.14±0.3ms         612±40μs     0.54  replace.FillNa.time_replace(True)
-        7.22±2ms       1.61±0.1ms     0.22  replace.FillNa.time_fillna_df(True)
-        8.97±2ms       2.00±0.2ms     0.22  replace.FillNa.time_fillna_limit(False)
-      5.47±0.2ms         639±30μs     0.12  replace.FillNa.time_fillna_limit(True)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.

```
</Details>

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -434,40 +434,56 @@ def fillna(
...
+        # TODO: Handle inf_as_na, we need to get option and pass to cython funcs
         inplace = validate_bool_kwarg(inplace, "inplace")
-
-        mask = isna(self.values)
-        mask, noop = validate_putmask(self.values, mask)
-
-        if limit is not None:
-            limit = libalgos.validate_limit(None, limit=limit)
-            mask[mask.cumsum(self.ndim - 1) > limit] = False
+        arr = self if inplace else self.copy()
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:439]: pls call this `blk`, as `arr` usually indicates ndarray or EA

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -434,40 +434,56 @@ def fillna(
...
+        limit = libalgos.validate_limit(
+            len(self) if self.ndim == 1 else self.shape[1], limit=limit
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:441]: self.shape[-1]?

File: pandas/core/internals/blocks.py [hunk 3]
```diff
@@ -434,40 +434,56 @@ def fillna(
...
+                )
         else:
-            # operate column-by-column
-            return self.split_and_operate(
-                type(self).fillna, value, limit=limit, inplace=inplace, downcast=None
-            )
+            # TODO: This seems to work for EAS, verify it does
+            return [
+                self.make_block_same_class(
+                    values=self.values.fillna(value=value, limit=limit)
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:484]: you'll probably need to make NDArrayBackedExtensionArray.fillna use the new cython function(s)

File: pandas/core/internals/blocks.py [hunk 4]
```diff
@@ -434,40 +434,56 @@ def fillna(
...
+            return [arr]
+
+        if not self.is_extension:
+            if self._can_hold_element(value):
+                if self.ndim == 1:
+                    if is_list_like(value):
+                        # TODO: Verify EA case
+                        if is_extension_array_dtype(value):
+                            mask = value.isna()
+                            value = np.asarray(value[mask], dtype=object)
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:454]: going straight to object may be unnecessarily heavy for e.g. Categorical `value`.

File: pandas/core/internals/blocks.py [hunk 5]
```diff
@@ -434,40 +434,56 @@ def fillna(
...
+                            libalgos.fillna1d_multi_values(
+                                arr.values[mask], value=value, limit=limit
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:456]: this is going to fill arr.values[mask] inplace, but i dont think arr.values[mask] shares data with arr.values?

**Comment by jbrockmendel** [comment]
My main advice is to not worry about the Block stuff for now, just get the cython code working and write tests for it directly with ndarrays (and NDArrayBackedExtensionArray potentiall)

**Comment by lithomas1** [comment]
@jbrockmendel Is it OK to do the NDArrayBacked.fillna changes in a followup, since this PR is already getting somewhat large?

**Comment by jbrockmendel** [comment]
Good idea, yes.

**Comment by mroeschke** [comment]
Appears this PR has been dormant for a while closing to clear the queue. Feel free to reopen when you have time to work on this PR.

**Pull request closed without merging**